### ISSUE TITLE
Add Barcode scanner component

### ISSUE THREAD

**Issue by maks** [Jasonette/Jasonette#10]
There are a number of uses for barcode scanning, especially 2D QR codes.

Initial use case would be to allow the Jason app to scan QR codes that point to jasonette json urls instead of requiring a user to input them (eg. as done by https://expo.io/)

It should be relatively easy to implement as there appears to be good support [on iOS](https://stackoverflow.com/a/19356282/85472) and on [Android](https://github.com/dm77/barcodescanner)

**Comment by drwasho** [comment]
Yes please

**Comment by arifrifai** [comment]
There is javascript porting of ZXing Java library. https://github.com/schmich/instascan

**Comment by maks** [comment]
@arifrifai yes thats a good point, its possible already using that via a webview, but that goes for alot of Jasonette functionality too. I think per @gliechtenstein view of having Jasonette be a "native" framework, it makes sense to use native support (via native code libraries and camera access APIs) for this.

**Comment by gliechtenstein** [comment]
# First Draft

Barcode scanning is an asynchronous process therefore we should implement it in the `action + service` pattern we have established at https://github.com/Jasonette/Jasonette/issues/6#issuecomment-325038968

Here's the first draft:

1. To use `$vision`, the `body.background` must be a `camera` type.
2. `$vision.ready` event gets triggered when the camera is ready.
3. We can trigger `$vision.scan` to start scanning for barcodes/qrcodes
4. The `$vision.scan` is asynchronous so it immediately goes on to its success callback after it kicks off the scanning service
5. The scanning service triggers a `$vision.onscan` event with the extracted metadata. We can use it to do whatever.

## Syntax

- Action
    - $vision.scan
- Events
    - $vision.onscan
    - $vision.ready

## Example

```
{
  "$jason": {
    "head": {
      "templates": {
        "body": {
          "style": {
            "background": {
              "type": "camera"
            }
          }
        }
      },
      "actions": {
        "$load": {
          "type": "$render"
        },
        "$vision.ready": {
          "type": "$vision.scan",
          "options": {
            "type": "qrcode"
          }
        },
        "$vision.onscan": {
          "type": "$util.alert",
          "options": {
            "title": "Scanned content",
            "description": "{{$jason.content}}"
           }
        }
      }
    }
  }
}
```

Here's what's going on above:

1. The `$load` triggers render, which draws a camera background
2. When the camera is ready, it triggers the `$vision.ready` event
3. This is when we can call the `$vision.scan` action (This can be called manually as well, but to automatically start scanning, we need to attach it to the `$vision.scan` event instead of `$load`)
4. `$vision.scan` returns immediately and the action call chain ends there. But the `$vision` service has been kicked off so it's constantly scanning
5. When the camera recognizes a code, it triggers `$vision.onscan` with the relevant payload.
6. We can use the content and do whatever we want, such as rendering it to the screen or doing an href or calling another action.

**Comment by gliechtenstein** [comment]
# Final Draft

There have been a bit of a change to simplify things. We could add more later if we want but for now have limited the attributes sent and received.

As of now this feature is available on develop branch of both iOS and Android:

- iOS: https://github.com/Jasonette/JASONETTE-iOS/commit/dac65c3b70367d9a417d3fed2e5d9c1fdae53c47
- Android: https://github.com/Jasonette/JASONETTE-Android/commit/533a9c806db0dde662a719e9d4e2bf43dde6a27b

Here's an example:

```
{
  "$jason": {
    "head": {
      "templates": {
        "body": {
          "style": {
            "background": {
              "type": "camera"
            }
          }
        }
      },
      "actions": {
        "$load": {
          "type": "$render"
        },
        "$vision.ready": {
          "type": "$vision.scan"
        },
        "$vision.onscan": {
          "type": "$util.alert",
          "options": {
            "title": "Scanned content",
            "description": "{{$jason.type}} {{$jason.content}}"
           }
        }
      }
    }
  }
}
```

1. `$vision.scan` doesn't take any parameter. it just scans any machine readable code supported by iOS ([list here](https://developer.apple.com/documentation/avfoundation/avmetadataobjecttype?language=objc)) and Android([list here](https://developers.google.com/android/reference/com/google/android/gms/vision/barcode/Barcode#constants))
2. As a callback, `$vision.onscan` only returns 2 attributes: `type` and `content` in the following format:
    ```
    [iOS]

    {
      "$jason": {
        "type": "org.iso.QRCode",
        "content": "Hello World"
       }
     }
    ```

    ```
    [Android]

    {
      "$jason": {
        "type": 256,
        "content": "Hello World"
       }
     }
    ```


3. `content` is obviously the raw string represented by the code
4. `type` represents the format of the scanned code. Right now iOS and Android have different ways of describing them and I have decided not to touch this. So for example when scanning a QR Code, iOS will return `"org.iso.QRCode"` whereas Android will return `256` (number). It is the developer's job to handle these using template expressions. But in most cases we don't really need to care about the type since all we care about is the content.

# Examples

Here are two examples using this API:

- Inline Detection and Rendering: https://jasonbase.com/things/PMj7 (any content can be encoded)
- detect link and $href: https://jasonbase.com/things/rZ0b (Be sure to scan a qrcode that encodes a link)
